Browser providers

NathanW2 opened the pull request:
The docs say that you can add a new QgsDataItemProvider in order to show it in the browser tree however that isn't the case as there is no signals and no way to update the tree after it is created.  

This PR adds a new signal and connects it to the QgsBrowserModel so new providers can be added and removed from plugins using the following

```
c = CustomProvider()
r = QgsApplication.dataItemProviderRegistry()
r.addProvider(c)
```

signedav committed:
setting of reload interval in featurelistcombo

cherry-picket from 13d25862630582f29bdc507bba48db3af7ba2aaa

signedav committed:
set value to null when pressed lineedit-clearbutton

cherry-picked from 713192f7106214588561945225628059796acc08

signedav committed:
do not overwrite text entry on editing and model reloadings

cherry-picked from 43f77fda5834d3d52707814f9d4019c7e8533c69

signedav committed:
remove setReloadTime because it's not used anymore

cherry-picked from 8a226671f57ee134fe6c348ef506a09295f1873e

signedav committed:
use lambda to set value to null

cherry-picked from fbbb32000e62d17c95b4c9649147259ed7363e07

signedav committed:
selectAll if requested on mouse release

because otherwise the mouse event makes the selectAll useless if it's done on focusEvent

cherry-picked from d3c861e744cefd4657e8a4d4219f634a8cd51d58

m-kuhn committed:
Ignore FID field when merging vector layers to geopackage

fuzzysolutions committed:
Minor changes to speed up PointstoPaths algorithm

# 1: Replaced time-consuming call on all attributes by a more specific request - responsible for about 90% of running time in my tests.
# 2: Minor speedup by calling items() before the for loop (1s in a large layer).

(cherry picked from commit 6c99cbc1ba341c08c192d763cfb67b82c8da5d49)

NathanW2 committed:
FIX - Add and connect signals to allow providers to be added to browser from plugins

signedav committed:
pass nullRepresentation to QgsFilterLineEdit

cherry-picked from e37f1d11272a3170b2335ea67c5b4d24a54594a1

NathanW2 commented:
This has a crash on close because of things going out of scope in Python as the registry doesn't take ownership.  Should it?

nyalldawson commented on src/core/qgsbrowsermodel.cpp:
```diff
@@ -380,6 +347,44 @@ void QgsBrowserModel::reload()
+void QgsBrowserModel::removeProvider( QgsDataItemProvider *provider )
+{
+  QgsDataItem *rootItem = mProviderMap[provider->name()];
+  removeRootItem( rootItem );
```
better check rootItem isn't null here

nyalldawson commented on src/core/qgsbrowsermodel.h:
```diff
@@ -162,6 +164,15 @@ class CORE_EXPORT QgsBrowserModel : public QAbstractItemModel
     void reload();
 
+    /**
+     * Add a new provider to the model.
```
@param /@return are bad :-1: Just go:

    * Adds a new \a provider to the model
    * \since QGIS 3.4

Also dox for removeProvider

nyalldawson commented on src/core/qgsdataitemproviderregistry.h:
```diff
@@ -55,6 +57,20 @@ class CORE_EXPORT QgsDataItemProviderRegistry
+
+    /**
+     * Fired when a new provider has been added to the registery
+     * @param provider The new data provider
```
\param or use the Qt style \a approach 

Also needs \since

nyalldawson commented:
> This has a crash on close because of things going out of scope in Python as the registry doesn't take ownership. Should it?

addProvider is marked as SIP_TRANSFER. Is that not working?

signedav committed:
remove connection to lineeditcleared, because of already correct null value implementation there

cherry-picked from 7939cfb5570e771004bf8ce9f034d450f99f6bce

signedav committed:
tests for switching between values

especially setting the NULL value and checking if it's selected
checking if the clear button is activ when it should and inactive when it should

git cherry-pick 2e01811aa789a595120a691ffc6a510b87e81047

signedav committed:
controll clear button by edit-state

because it has to appear on edit-mode, otherwise not

cherry-picked from f0b0aea3b223b1af4a61e1ce01513d1e8cbdf6d3

signedav committed:
Test if NULL is selected

cherry-picked from 5b4aed3aa2e6f40785cba9f67d887b729cdd0c9e

signedav committed:
removed extra ;

cherry-picked from a4bb9058b444c0fe3a7bab91fb2e33179c3498b9

signedav committed:
handle EnableChange event in QgsFilterLineEdit

cherry-picked from 9768fdd0f3a302ea058ba7d1ab4fc1a205750795

m-kuhn committed:
Allow using gpkg in processing tests

m-kuhn committed:
Add merge vector layer test for gpkg

m-kuhn committed:
Fix faulty alg name in test

NathanW2 committed:
Add null check

Gustry committed:
remove translations for expression group to fix the UI

Gustry committed:
add missing expression groups

rldhont committed:
[Bugfix][Server] Use Project selection color in GetPrint 3.4

jef-n committed:
Release of 3.4 (Madeira)

3nids committed:
add missing GUI_EXPORT macro

jef-n committed:
add islandic translator (followup f4e6477a)

(cherry picked from commit f3b0b257e424557d49a6e6f77c4339d3be7afc8e)

jef-n committed:
scripts/tsstat.pl: remove unused site mode and add note to doc/TRANSLATORS

(cherry picked from commit 012bd99e110c221f7cfc3b9148e7cceb2dea443b)

nyalldawson committed:
Follow up status bar font fixes, avoid too small font on Windows

(cherry picked from commit c9206884372e25decb354706728935cef5765e34)

nyalldawson committed:
Avoid useless tray icon on Windows 7

Fixes #20247

(cherry picked from commit 908a058a3a88947498be44e8bba1ce2afb12981e)

nyalldawson committed:
[processing] Fix exceptions due to closed algorithm dialogs when
opened from toolbox and closed before task completes

(cherry picked from commit 0d5a8ddd8e387219721ff24cb8299960e4f3d334)

nyalldawson committed:
Capitalization

(cherry picked from commit 1dd36f132d1fbf04b0aa265157d5547fe8aa9140)

nyalldawson committed:
Avoid creating temporary project properties and options dialogs
for locator filter

This is expensive and results in a lengthy delay when first
using the locator in a QGIS session.

Replace with an alternative approach which doesn't rely on
creation of temporary dialogs, yet should still ensure that
these dialogs don't go out of sync with the locator.

(cherry picked from commit 0035b7c36965b144dc5e6c1b6684c4edcc9286e7)

nyalldawson committed:
Fix clazy warnings

(cherry picked from commit 3c3fcf50a9a95d2e818eddbc241616668efef3eb)

nyalldawson committed:
More clazy optimisations

(cherry picked from commit d4cf8cf7de995f0ba5a9ef4a01276de8400e7fb2)

wonder-sk commented on src/app/qgisapp.cpp:
```diff
@@ -1033,6 +1033,8 @@ QgisApp::QgisApp( QSplashScreen *splash, bool restorePlugins, bool skipVersionCh
   connect( mBrowserWidget, &QgsBrowserDockWidget::openFile, this, &QgisApp::openFile );
   connect( mBrowserWidget, &QgsBrowserDockWidget::handleDropUriList, this, &QgisApp::handleDropUriList );
+  connect( QgsApplication::dataItemProviderRegistry(), &QgsDataItemProviderRegistry::providerAdded, mBrowserModel, &QgsBrowserModel::addProvider );
+  connect( QgsApplication::dataItemProviderRegistry(), &QgsDataItemProviderRegistry::providerWillBeRemoved, mBrowserModel, &QgsBrowserModel::removeProvider );
```
I think rather than just connecting to registry from outside of the model, QgsBrowserModel should connect to it by itself in the constructor. So it will work with all browser models automatically...

wonder-sk commented on src/core/qgsbrowsermodel.cpp:
```diff
@@ -123,42 +123,9 @@ void QgsBrowserModel::addRootItems()
-    if ( providerGroup.size() > 1 )
-    {
-      std::sort( providerGroup.begin(), providerGroup.end(), cmpByDataItemName_ );
-    }
```
It seems the sorting got dropped - is that intentional?

nyalldawson replied:
I believe it's redundant now - sorting is done by the QgsBrowserProxyModel by item sort key.

wonder-sk commented on src/core/qgsbrowsermodel.h:
```diff
@@ -162,6 +164,15 @@ class CORE_EXPORT QgsBrowserModel : public QAbstractItemModel
+     */
+    bool addProvider( QgsDataItemProvider *provider );
+
+    void removeProvider( QgsDataItemProvider *provider );
```
these two should be private slots ideally

wonder-sk reviewed (commented):
Regarding the crash you're getting - how does your code look like exactly? do you keep reference to the provider in python while the provider is active?

elpaso committed:
Fix geometry precision input in vector layer properties and not-dot locales

Fix #20255 - Impossible to enter a decimal value for the geometry precision in French (comma as decimal separator)

Backported from master

m-kuhn committed:
Hide geometry validation dock

Fix #20245 https://issues.qgis.org/issues/20245

elpaso committed:
Show an empty string if precision is 0 or not valid

Backported from master

m-kuhn committed:
Fix deadlock in geometry validation

Fix a deadlock in geometry validation. This happens when the mainthread "waits for finished" of a checker thread.

What happened in this case was, that the main thread canceled the feedback and waited for any jobs to finish.
If a job was waiting for a feature source (or something else to be executed on the main thread) this resulted in a mighty deadlock.

What we do here is that we regularly check if we ought to cancel while waiting for our slot on the main thread and can bail out if we should cancel before the main thread gets around to take care of our function.
The difference is, we still execute the code on the background thread and make sure that the main thread is not doing anything during this time to avoid working on the same data structures in parallel.

m-kuhn committed:
Silence warning

m-kuhn committed:
Allow calling getFeatureSource without feedback

bjornharrtell committed:
[oracle] Prefetch memory variable changed to 32-bit and default raised to 8MB

Speeds up oracle table access by orders of magnitude

(cherry picked from commit c3e7632ce8c8370cc7b6941e3be523844d8bf705)

ricardogsilva committed:
vector layer properties now checks if config widget factory supports layer

previously the factory would always be used, even if it did not
support the active layer

fixes #20240

(cherry picked from commit a9f7f3542e2424b0b8cbd500c077ea648d6f24f6)

nyalldawson committed:
[processing] Fix crash when right clicking toolbox outside of an item

Fixes #20264

(cherry picked from commit a873289da7846b6761fe40733d5e7407cce7684d)

nyalldawson committed:
Fix broken handle bad layers dialog

Partially revert af0bae67f and ea2cc365

Fixes #20265

(cherry picked from commit 0d7f59865b072c29f853513eb183c674585981ba)

Commit:
update mac bundling for new Qt frameworks

nyalldawson committed:
Avoid crash in style model when adding symbols with a new tag to library

Fixes #20256

(cherry picked from commit f6aecd152d367ed0a4cf661c3c5c8c1850f608af)

nyalldawson committed:
[browser] Fix some items which are marked as "Collapse" capability
are incorrectly expanded on startup

(cherry picked from commit 5ebb9745694de324f6506d83536f403f6b001b01)

nirvn committed:
Prevent flooding (and app freeze) of items added to the message bar

m-kuhn committed:
Unify parameter names

m-kuhn committed:
Use decodeUri instead of string parsing

m-kuhn committed:
More comprehensive API docs

m-kuhn committed:
Merge pull request #8373 from m-kuhn/hideGeometryValidationBackport

Hide geometry validation dock

m-kuhn committed:
Make doxygen a bit happier

m-kuhn committed:
Fix test for merging layers to gpkg

m-kuhn committed:
Add comment for semi-broken control file

m-kuhn committed:
Merge pull request #8375 from m-kuhn/deadlockInGeometryValidation

Backport of fix deadlock in geometry validation

m-kuhn committed:
Add missing param doc

m-kuhn committed:
Skip path checking

m-kuhn committed:
Parameter name harmonization

nyalldawson committed:
Fix a (harmless) build warning

(cherry picked from commit 2c278f31184084e9a39ad811e90af069495c3994)

nyalldawson committed:
Fix another warning

(cherry picked from commit 3a18ba6b243c23f3dd7bfbc7268680e8875318bf)

nyalldawson committed:
Fix inefficient calls to QgsFeature.attributes()[idx]

Replace with just f[idx]. Calling QgsFeature.attributes()
allocates a list of all attributes, which is inefficient
when only a single attribute value is needed.

(cherry picked from commit 85efc77bb4441c645868931eeb2fe8476031a645)

elpaso committed:
Merge pull request #8379 from elpaso/backport-3_4

Backport: Fix geometry precision input in vector layer properties and not-dot

nyalldawson committed:
Fix some inefficient python dictionary iteration

(cherry picked from commit f3e9aaf79a9282b28a605abd0dadaab9951050c8)

nyalldawson committed:
Add more debugging to auth db opening

(cherry picked from commit ac781fac50ff0d48f5c267a64567fc0bb48548f1)

3nids committed:
native/CMakeLists.txt cleanup (#8380)

nyalldawson committed:
Correctly cleanup auth db connection on thread finished

(doesn't fix the invalid connection issue!)

(cherry picked from commit 368286fed0ca2fe752fef083d97235a1eda444f9)

nyalldawson committed:
Fix authentication database connections

Fixes the authentication database cannot be opened in some circumstances.
We need to ensure that the pooled database connection is removed
immediately on thread finalisation and cannot defer this until
the main thread event loop runs.

Fixes #20262

(cherry picked from commit a392a160e6d7fdc70e3da857dc1aedb601a70007)

nyalldawson committed:
[mssql] Fix database connection cleanup on thread exit

(cherry picked from commit a5ff6db24066642e2b5952b22bfe069013ec9a0b)

nyalldawson committed:
[browser] Don't try to open links within the browser properties widget

Instead open them externally. Fixes clicking on the path link
within metadata tries to open layer files instead the text
widget instead of showing them in the desktop file browser.

(cherry picked from commit ab4601820cf5232e1ed76f74432bc2000935276c)

nyalldawson committed:
[browser] Remove redundant name/uri/provider labels from top
of layer properties widget

All this information is already shown in the scrollable metadata
section below that, so we're wasting precious space in this
valuable widget with a lot of duplicate information

(cherry picked from commit 7f332fb7a635ebaea64dfc787926c61b98338b54)

nyalldawson committed:
[browser] Unify layer name in properties for raster/vector layers

(cherry picked from commit 620643e2543b76deb5bc209d765f9356dfbe6534)

m-kuhn committed:
Case insensitive file extension check

m-kuhn committed:
Add test for QgsVectorLayerExporter with RegeneratePrimaryKey

m-kuhn committed:
Regenerate fids for "join attributes by location"

Fix #19708

m-kuhn committed:
Regenerate fids for "points along geometry"

Fix #19998

m-kuhn committed:
Regenerate fids for "Join attributes by field value"

m-kuhn committed:
Regenerate fids for "join attributes by location", pt2

m-kuhn committed:
Regenerate fids for "multipart to singlepart algorithm"

rldhont committed:
Merge pull request #8349 from rldhont/bugfix-server-getprint-color-selection-3

[Bugfix][Server] Use Project selection color in GetPrint 3.4

elpaso committed:
[wms] Do not pass null QStrings to QgsWmsProvider::setQueryItem

Fix #20271 - WMS is not displayed in QGIS 3.4.0

With test

Cherry-picked from 94aa628e3e94e5998ba762810e19be104c0f820b

elpaso committed:
Skip kw check

m-kuhn committed:
Make sip a bit happier

elpaso committed:
Merge pull request #8389 from elpaso/bugfix-20271-wms-null-styles-backport

[wms] Do not pass null QStrings to QgsWmsProvider::setQueryItem

jef-n committed:
german translation update

nyalldawson committed:
Fix build warning

(cherry picked from commit 41c31b44ba81649f76e048185a18c4f6073addd3)

nyalldawson committed:
Fix crash on opening options on debug builds where open CL is not enabled

Fixes #20295

(cherry picked from commit 15435ad56e24bd8dcfdee1556eb25e42fa31bcb3)

nyalldawson committed:
[processing] Fix incorrect html used when pushing console command feedback to dialog

And make color less dominant (not bright blue!). We want other more critical
messages the be more attention grabbing then the noisy console outputs.

(cherry picked from commit 484895e2673ad854e8d18f61d44ee5dd62f4ea64)

nyalldawson committed:
[processing][grass] Highlight GRASS error messages and warnings in log

And catch segmentation faults from GRASS commands and show a nice
helper text advising user that QGIS isn't at fault here ;)

Also add a tip to try disabling the v.external option if that setting
is enabled and a GRASS segfault occurs

(cherry picked from commit 676238b28fad1c453c1f51addd71065537e85b75)

nyalldawson committed:
[processing][GRASS] Give some more feedback when trying to use
v.external for reading inputs, and if we can't, advise why not

(cherry picked from commit 82ac04d6f77c6ac4c0699cc6e7302625bc704a39)

nyalldawson committed:
[processing][GRASS] Correctly handle input vector layers with |layername= param

Fixes #20277

(cherry picked from commit f65c845f86a9742602b72584b34b6ae63ed30dd7)

nyalldawson committed:
Hidpi friendly welcome page sizes

(cherry picked from commit 26c3dd30a9f736187debc95d668ee905ef2dcc12)

Commit:
don't run GRASS commands in shell, that's not working

nyalldawson committed:
Hidpi friendly message bar icon sizes

(cherry picked from commit e83aa5a200ebad78916f26ab291bcc5915291863)

m-kuhn committed:
Merge pull request #8392 from m-kuhn/mergeDuplicateFidBackport

Processing: GeoPackage output, regenerate fid when there's a risk of duplicates

m-kuhn committed:
Merge pull request #8374 from signedav/bugfix_not_deleting_relation-3_4

Backport More intuitive relation reference widget text filter

nyalldawson committed:
Tighten scope of connection

(cherry picked from commit c782f2782016f9be0f127de25a42e48e82ead378)

nyalldawson committed:
Fix memory leak

(cherry picked from commit 3195c82ec1347dc57748f6acaa648868138b4b11)

nyalldawson committed:
Fix build

rouault committed:
[BUGFIX] [OGR provider] Make filter by id(s) requests work again on OSM datasets (fixes #20308)

jef-n committed:
verify that projecta actually delivers a changelog

(cherry picked from commit a826c5644549216491e2e101a5809c64c5193421)

jef-n committed:
fix news

(cherry picked from commit 1151639f9d928ddb0ee8e482ef5a61a636b90ffd)

3nids committed:
fix qgsDoubleString printing -0 (#8363) (#8398)

nyalldawson committed:
Fix crash on Windows on idle/usb notifications

Fixes #20312, #20319, #20312, #20307, #20299, #20297, #20280
(maybe others)

(cherry picked from commit fd5a1c82ec212c58c446ccb13e537484d44b6862)

nyalldawson committed:
Fix potential crash on exit

jef-n committed:
german translation update

NathanW2 commented:
Hello stalebot :)

On Fri, Nov 16, 2018 at 9:56 AM stale[bot] <notifications@github.com> wrote:

> The QGIS project highly values your contribution and would love to see
> this work merged! Unfortunately this PR has not had any activity in the
> last 14 days and is being automatically marked as "stale". If you think
> this pull request should be merged, please check
>
>    -
>
>    that all unit tests are passing
>    -
>
>    that all comments by reviewers have been addressed
>    -
>
>    that there is enough information for reviewers, in particular
>    -
>
>       link to any issues which this pull request fixes
>       -
>
>       add a description of workflows which this pull request fixes
>       -
>
>       add screenshots if applicable
>       -
>
>    that you have written unit tests where possible
>    In case you should have any uncertainty, please leave a comment and we
>    will be happy to help you proceed with this pull request.
>    If there is no further activity on this pull request, it will be
>    closed in a week.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/8315#issuecomment-439233569>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3KUN-WZLqzTSGnQARBkb9XVIdlseks5uvf8_gaJpZM4X5hnY>
> .
>

nyalldawson commented:
Yes -- don't let this one die please!

NathanW2 commented:
Will get it fixed up after foss4g, or during depending on time :)


On Fri, Nov 16, 2018 at 10:31 AM Nyall Dawson <notifications@github.com>
wrote:

> Yes -- don't let this one die please!
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/8315#issuecomment-439239955>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3DmnBWr3wSo5NKWz8-HCt1VNVAGkks5uvgdvgaJpZM4X5hnY>
> .
>

nyalldawson commented:
> Will get it fixed up after foss4g, or during depending on time :)

Nope, no star realms until this is merged.

NathanW2 commented:
Pretty sure we are both losing from that deal

On Fri, Nov 16, 2018 at 10:59 AM Nyall Dawson <notifications@github.com>
wrote:

> Will get it fixed up after foss4g, or during depending on time :)
>
> Nope, no star realms until this is merged.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/8315#issuecomment-439244830>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3Bs9X_uArh2HP4kdB3rAtoMlZd7uks5uvg3ygaJpZM4X5hnY>
> .
>

nyalldawson commented:
While I agree with bringing this work back - I think this branch is well and truely in a broken state :stuck_out_tongue:

NathanW2 commented:
I suspect so, very much.  I will bring it back and fix it up.  Should also
be safe for a backport hopefully.

On Fri, Sep 6, 2019 at 10:31 AM Nyall Dawson <notifications@github.com>
wrote:

> While I agree with bringing this work back - I think this branch is well
> and truely in a broken state 😛
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/8315?email_source=notifications&email_token=AAC5FXBGETJMSBVVTSP7GA3QIGQEZA5CNFSM4F7GDHMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6BLDYY#issuecomment-528658915>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC5FXHU7P7MQOSDEDU74E3QIGQEZANCNFSM4F7GDHMA>
> .
>
